File: day01/solution_part2.exs
{:ok, input} = File.read("input.txt")
lines = String.split(input, "\n")

IO.inspect(lines)
IO.puts("lines: #{inspect(length(lines))}")

Enum.reduce(lines, %{previous_lines: [], times_increased: 0}, fn curr_str,
                                                                 %{
                                                                   previous_lines: previous_lines,
                                                                   times_increased:
                                                                     times_increased
                                                                 } ->
  {curr_num, ""} = Integer.parse(curr_str)

  case previous_lines do
    [prev1, prev2, prev3] ->
      prev3_sum = prev1 + prev2 + prev3
      curr3_sum = prev2 + prev3 + curr_num

      %{
        previous_lines: [prev2, prev3, curr_num],
        times_increased: if(curr3_sum > prev3_sum, do: times_increased + 1, else: times_increased)
      }

    _ ->
      %{
        previous_lines: previous_lines ++ [curr_num],
        times_increased: times_increased
      }
  end
end)
|> IO.inspect()

File: haskell/day01/solution_part2.hs
module Main (main) where

import Text.Printf (printf)

data TimesIncreased = TimesIncreased
  { previous :: [Integer],
    timesIncreased :: Integer
  }
  deriving (Show)

countTimesIncreased :: TimesIncreased -> String -> TimesIncreased
countTimesIncreased TimesIncreased {previous = [prev1, prev2, prev3], timesIncreased = timesIncreased} currStr =
  TimesIncreased {previous = [prev2, prev3, curr], timesIncreased = newTimesIncreased}
  where
    curr = read currStr :: Integer
    prev3Sum = prev1 + prev2 + prev3
    curr3Sum = prev2 + prev3 + curr
    newTimesIncreased = if curr3Sum > prev3Sum then timesIncreased + 1 else timesIncreased
countTimesIncreased TimesIncreased {previous = previous, timesIncreased = timesIncreased} currStr =
  TimesIncreased {previous = previous ++ [curr], timesIncreased = timesIncreased}
  where
    curr = read currStr :: Integer

main :: IO ()
main = do
  _lines <- fmap lines (readFile "input.txt")
  printf "num lines: %d\n" (length _lines)
  print $ foldl countTimesIncreased (TimesIncreased {previous = [], timesIncreased = 0}) _lines
